classes/form/enrolkey_signup_form.php

```diff
@@ -78,49 +78,66 @@ public function definition() {
      *         or an empty array if everything is OK (true allowed for backwards compatibility too).
      */
     public function validation($data, $files) {
-        global $DB;
         $errors = parent::validation($data, $files);
 
-        $enrolplugin = enrol_get_plugin('self');
+        $signuptoken = $data['signup_token'];
 
-        $token = $data['signup_token'];
+        if ($signuptoken !== '') {
+            // For any case where the token is populated, perform a lookup.
+            $tokenisvalid = $this->check_database_for_enrolkey($signuptoken);
 
-        if (!empty($token)) {
-            $selfenrolinstance = false;
+            if ($tokenisvalid === false) {
+                $errors['signup_token'] = get_string('signup_token_invalid', 'auth_enrolkey');
+            }
 
-            $instances = $DB->get_records('enrol', array('password' => $token, 'enrol' => 'self'));
+        } else {
+            // The form submission is an empty string, double check if the token is required.
+            if ($this->signup_token_required()) {
+                $errors['signup_token'] = get_string('signup_missing', 'auth_enrolkey');
+            }
+        }
 
-            // There may be more than one enrolment instance configured with various dates to check against.
-            foreach ($instances as $instance) {
-                // There may be things that prevent self enrol, such as requiring a capability, or full course.
-                // This should not be a blocker to account creation. The creation should pass, then report the error.
-                if ($instance->status == ENROL_INSTANCE_ENABLED) {
-                    $selfenrolinstance = true;
-                }
+        return $errors;
+    }
+
+    /**
+     * Checks the enrolment records for any matching self enrolment key.
+     *
+     * @param $token Returns true on success. False on failure.
+     * @return bool
+     */
+    private function check_database_for_enrolkey($token) {
```

Comment:
I think this should probably be renamed to check_database_for_signuptoken as the rest of the class all references it as signuptoken

Reply:
Cheers, good point.